db/snapshot/backup_task.cc

```diff
@@ -82,6 +85,9 @@ void backup_task_impl::do_backup() {
         }).get();
         impl::_as.check();
     }
+    if (ex) {
+        std::rethrow_exception(ex);
+    }
```

Comment:
This looks wrong. By the time the above loop terminates changes that _any_ of background uploading fibers steps on exception are tiny. So this `if` will be evaluated to false and on-stack destructors will start being called. 

And this leads to another problem -- the `ex` is declared _after_ `wait`, so it will be destructed first, then stack destroying will pause waiting for the bacground gate to close. If by that time any background fiber resolves into exception, it will effectively reference destructed `ex`

Reply:
thank you Pavel for pointing out this. i should have 

- closed the gate first, and then check the captured `ex`.
- ensure that `ex` outlive `uploads` gate

anyway, this would probably need to increase the indent level. so i am reverting to the coroutinized approach.